### PR TITLE
Add jovyan user

### PR DESCRIPTION
**Pull request by paskino**
This adds the `jovyan` user to the docker image

### CODE REVIEW

File: docker/Dockerfile [hunk 1]
```diff
@@ -112,3 +132,22 @@ CMD ["/usr/local/bin/service.sh"]
...
+ARG NB_UID="1000"
+ARG NB_GID="100"
+# Copy a script that we will use to correct permissions after running certain commands
+COPY fix-permissions /usr/local/bin/fix-permissions
+RUN chmod a+rx /usr/local/bin/fix-permissions
+# Create NB_USER with name jovyan user with UID=1000 and in the 'users' group
+# and make sure these dirs are writable by the `users` group.
+RUN echo "auth requisite pam_deny.so" >> /etc/pam.d/su && \
+    sed -i.bak -e 's/^%admin/#%admin/' /etc/sudoers && \
+    sed -i.bak -e 's/^%sudo/#%sudo/' /etc/sudoers && \
```
Comment (KrisThielemans) [inline docker/Dockerfile:146]: no clue what this does... I suppose add a comment

**Review by KrisThielemans — COMMENTED** [review]
looks mostly ok, although I don't know why we're doing this of course. Add some comments once we're sure.

Doesn't look like we're running `fix-permissions` at the moment, but it is of course equivalent to what we're doing, and hence will be slow.

**Comment by paskino** [comment]
Does it make any difference when we create the jovyan user?

File: docker/Dockerfile (hunk 1 above)
Comment (KrisThielemans) [reply docker/Dockerfile:146]: some to be commenting out some sudoers. I'd have to check what the pam stuff does, but will leave that for someone else.

**Comment by KrisThielemans** [comment]
> Does it make any difference when we create the jovyan user?

not at all. I'm slightly more worried about the `/etc/passwd` manipulations etc. You must have got them from somewhere so I guess it's alright. (although I'm not sure if people really know how to handle docker/kubernetes file permissions properly, or indeed if it's possible...)

**Comment by KrisThielemans** [comment]
maybe clean-up history a bit here (squash merge would be ok, except that you squeezed in the CIL_demos!)

**Comment by paskino** [comment]
It all comes from https://github.com/jupyter/docker-stacks/blob/master/base-notebook/Dockerfile 

I'll merge once I'm happy with it and will remove the CIL-Demos.

**Comment by paskino** [comment]
Currently `gadgetron` complains

```
06-10 21:26:09.981 DEBUG [gadgetron_home.cpp:92] Executable path: "/opt/SIRF-SuperBuild/INSTALL/bin/gadgetron"

06-10 21:26:09.982 DEBUG [gadgetron_home.cpp:98] Gadgetron home: "/opt/SIRF-SuperBuild/INSTALL"

06-10 21:26:09.982 INFO [GadgetStreamController.cpp:48] Connection from 127.0.0.1:38828
06-10 21:26:09.982 INFO [GadgetStreamController.cpp:226] Found 1 readers
06-10 21:26:09.982 INFO [GadgetStreamController.cpp:227] Found 1 writers
06-10 21:26:09.982 INFO [GadgetStreamController.cpp:228] Found 3 gadgets
06-10 21:26:09.982 INFO [GadgetStreamController.cpp:244] --Found reader declaration
06-10 21:26:09.982 INFO [GadgetStreamController.cpp:245]   Reader dll: gadgetron_mricore
06-10 21:26:09.982 INFO [GadgetStreamController.cpp:246]   Reader class: GadgetIsmrmrdAcquisitionMessageReader
06-10 21:26:09.982 INFO [GadgetStreamController.cpp:247]   Reader slot: 1008
06-10 21:26:09.985 ERROR [GadgetStreamInterface.h:57] Failed to load DLL, Possible reasons:
06-10 21:26:09.985 ERROR [GadgetStreamInterface.h:58]    * Name of DLL is wrong in XML file
06-10 21:26:09.985 ERROR [GadgetStreamInterface.h:59]    * Path of DLL is not in your DLL search path (LD_LIBRARY_PAT
H on Unix)
06-10 21:26:09.986 ERROR [GadgetStreamInterface.h:60]    * Path of other DLLs that this DLL depends on is not in the
search path
```

I have 
```
$ echo $LD_LIBRARY_PATH
/opt/SIRF-SuperBuild/INSTALL/lib:/opt/SIRF-SuperBuild/INSTALL/lib64:/usr/local/nvidia/lib:/usr/local/nvidia/lib64:

$ echo $PATH
/opt/SIRF-SuperBuild/INSTALL/bin:/usr/local/nvidia/bin:/usr/local/cuda/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin
```

**Comment by paskino** [comment]
Found the trouble:

```

$ ldd /opt/SIRF-SuperBuild/INSTALL/lib/libgadgetron_mricore.so
        linux-vdso.so.1 (0x00007ffc32ffa000)
        libgadgetron_gadgetbase.so.3.17 => not found
        libgadgetron_toolbox_denoise.so.3.17 => not found
        libgadgetron_toolbox_hostutils.so.3.17 => not found
        libismrmrd.so.1.4 => not found
        libpthread.so.0 => /lib/x86_64-linux-gnu/libpthread.so.0 (0x00007f6ac5d71000)
        libACE-6.4.5.so => /usr/lib/libACE-6.4.5.so (0x00007f6ac59df000)
        libarmadillo.so.8 => /usr/lib/libarmadillo.so.8 (0x00007f6ac57d6000)
        libboost_filesystem.so.1.65.1 => /usr/lib/x86_64-linux-gnu/libboost_filesystem.so.1.65.1 (0x00007f6ac55bc000)
        libgadgetron_toolbox_cpuoperator.so.3.17 => not found
        libgadgetron_toolbox_mri_core.so.3.17 => not found
        libgadgetron_toolbox_image_analyze_io.so.3.17 => not found
        libgadgetron_toolbox_cpufft.so.3.17 => not found
        libgadgetron_toolbox_cpuklt.so.3.17 => not found
        libgadgetron_toolbox_cpucore_math.so.3.17 => not found
        libgadgetron_toolbox_log.so.3.17 => not found
        libstdc++.so.6 => /usr/lib/x86_64-linux-gnu/libstdc++.so.6 (0x00007f6ac5233000)
        libm.so.6 => /lib/x86_64-linux-gnu/libm.so.6 (0x00007f6ac4e95000)
        libgomp.so.1 => /usr/lib/x86_64-linux-gnu/libgomp.so.1 (0x00007f6ac4c66000)
        libgcc_s.so.1 => /lib/x86_64-linux-gnu/libgcc_s.so.1 (0x00007f6ac4a4e000)
        libc.so.6 => /lib/x86_64-linux-gnu/libc.so.6 (0x00007f6ac465d000)
        /lib64/ld-linux-x86-64.so.2 (0x00007f6ac643e000)
        libdl.so.2 => /lib/x86_64-linux-gnu/libdl.so.2 (0x00007f6ac4459000)
        librt.so.1 => /lib/x86_64-linux-gnu/librt.so.1 (0x00007f6ac4251000)
        libblas.so.3 => /usr/lib/x86_64-linux-gnu/libblas.so.3 (0x00007f6ac3c8f000)
        liblapack.so.3 => /usr/lib/x86_64-linux-gnu/liblapack.so.3 (0x00007f6ac33d0000)
        libarpack.so.2 => /usr/lib/x86_64-linux-gnu/libarpack.so.2 (0x00007f6ac3186000)
        libsuperlu.so.5 => /usr/lib/x86_64-linux-gnu/libsuperlu.so.5 (0x00007f6ac2f16000)
        libboost_system.so.1.65.1 => /usr/lib/x86_64-linux-gnu/libboost_system.so.1.65.1 (0x00007f6ac2d11000)
        libgfortran.so.4 => /usr/lib/x86_64-linux-gnu/libgfortran.so.4 (0x00007f6ac2932000)
        libquadmath.so.0 => /usr/lib/x86_64-linux-gnu/libquadmath.so.0 (0x00007f6ac26f2000)
```

**Comment by paskino** [comment]
That's not it: after sourcing the env file all dll are found

```
 source /opt/SIRF-SuperBuild/INSTALL/bin/env_sirf.sh
jovyan@e32ebb7cae2c:~$ echo $LD_LIBRARY_PATH
/opt/SIRF-SuperBuild/INSTALL/lib:/opt/SIRF-SuperBuild/INSTALL/lib64:/usr/local/nvidia/lib:/usr/local/nvidia/lib64:
jovyan@e32ebb7cae2c:~$ ldd /opt/SIRF-SuperBuild/INSTALL/lib/libgadgetron_mricore.so
        linux-vdso.so.1 (0x00007fff2e1da000)
        libgadgetron_gadgetbase.so.3.17 => /opt/SIRF-SuperBuild/INSTALL/lib/libgadgetron_gadgetbase.so.3.17 (0x00007f67ed414000)
        libgadgetron_toolbox_denoise.so.3.17 => /opt/SIRF-SuperBuild/INSTALL/lib/libgadgetron_toolbox_denoise.so.3.17 (0x00007f67ed1d5000)
        libgadgetron_toolbox_hostutils.so.3.17 => /opt/SIRF-SuperBuild/INSTALL/lib/libgadgetron_toolbox_hostutils.so.3.17 (0x00007f67ecfba000)
        libismrmrd.so.1.4 => /opt/SIRF-SuperBuild/INSTALL/lib/libismrmrd.so.1.4 (0x00007f67ecd35000)
```

**Comment by KrisThielemans** [comment]
> That's not it: after sourcing the env file all dll are found

so, does it run then after sourcing?

Is the env file read automatically by this user?

**Comment by paskino** [comment]
No, it doesn't work well yet.

**Comment by paskino** [comment]
@KrisThielemans and I found that Gadgetron's CMake found hdf5 in the conda directory which isn't in the `PATH` by default. Starting Gadgetron with `LD_LIBRARY_PATH=${CONDA_PREFIX}/lib:$LD_LIBRARY_PATH ./INSTALL/bin/gadgetron` does the trick.

https://github.com/paskino/SIRF-SuperBuild/blob/f753289d70d41e256327cbcd8a670f8b9faa8921/docker/service.sh#L45

I also modified the system wide jupyter as jupyterhub is not going to start `jupyter` rather `jupyterhub-singleuser`.

**Comment by paskino** [comment]
I'll close this PR as it is really a work-in-progress.

**Pull request closed without merging**